{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "snakemake-job-properties"
    ]
   },
   "outputs": [],
   "source": [
    "\n",
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "import sys; sys.path.extend(['/cluster/ggs_lab/mtparker/.conda/envs/snakemake6/lib/python3.10/site-packages', '/cluster/ggs_lab/mtparker/papers/splice_site_phylogenetics/phylo_assoc_mapping/rules/notebook_templates']); import pickle; snakemake = pickle.loads(b'\\x80\\x04\\x95Ff\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x10snakemake.script\\x94\\x8c\\tSnakemake\\x94\\x93\\x94)\\x81\\x94}\\x94(\\x8c\\x05input\\x94\\x8c\\x0csnakemake.io\\x94\\x8c\\nInputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8c3pssms_inferred/Hanseniaspora_guilliermondii.5ss.tsv\\x94\\x8c*pssms_inferred/Yarrowia_lipolytica.5ss.tsv\\x94\\x8c0pssms_inferred/Wickerhamiella_versatilis.5ss.tsv\\x94\\x8c1pssms_inferred/Nakaseomyces_bacillisporus.5ss.tsv\\x94\\x8c0pssms_inferred/Saccharomyces_pastorianus.5ss.tsv\\x94\\x8c,pssms_inferred/Wickerhamomyces_bovis.5ss.tsv\\x94\\x8c(pssms_inferred/Candida_tropcalis.5ss.tsv\\x94\\x8c+pssms_inferred/Candida_metapsilosis.5ss.tsv\\x94\\x8c(pssms_inferred/Yamadazyma_tenuis.5ss.tsv\\x94\\x8c\\'pssms_inferred/Pichia_terricola.5ss.tsv\\x94\\x8c+pssms_inferred/Hyphopichia_burtonii.5ss.tsv\\x94\\x8c0pssms_inferred/Metschnikowia_bicuspidata.5ss.tsv\\x94\\x8c/pssms_inferred/Metschnikowia_spJCM33374.5ss.tsv\\x94\\x8c-pssms_inferred/Zygotorulaspora_mrakii.5ss.tsv\\x94\\x8c$pssms_inferred/Candida_auris.5ss.tsv\\x94\\x8c0pssms_inferred/Schizosaccharomyces_pombe.5ss.tsv\\x94\\x8c-pssms_inferred/Alloascoidea_hylecoeti.5ss.tsv\\x94\\x8c-pssms_inferred/Tetrapisispora_blattae.5ss.tsv\\x94\\x8c,pssms_inferred/Kregervanrija_fluxuum.5ss.tsv\\x94\\x8c+pssms_inferred/Lipomyces_oligophaga.5ss.tsv\\x94\\x8c*pssms_inferred/Nadsonia_fulvescens.5ss.tsv\\x94\\x8c-pssms_inferred/Pichia_membranifaciens.5ss.tsv\\x94\\x8c\\'pssms_inferred/Candida_gorgasii.5ss.tsv\\x94\\x8c,pssms_inferred/Teunomyces_gatunensis.5ss.tsv\\x94\\x8c-pssms_inferred/Barnettozyma_salicaria.5ss.tsv\\x94\\x8c.pssms_inferred/Saccharomycodes_ludwigi.5ss.tsv\\x94\\x8c)pssms_inferred/Dipodascus_albidus.5ss.tsv\\x94\\x8c.pssms_inferred/Kazachstania_naganishii.5ss.tsv\\x94\\x8c+pssms_inferred/Millerozyma_farinosa.5ss.tsv\\x94\\x8c/pssms_inferred/Kregervanrija_delftensis.5ss.tsv\\x94\\x8c1pssms_inferred/Brettanomyces_naardenensis.5ss.tsv\\x94\\x8c*pssms_inferred/Suhomyces_pyralidae.5ss.tsv\\x94\\x8c2pssms_inferred/Saccharomycopsis_fibuligera.5ss.tsv\\x94\\x8c*pssms_inferred/Debaryomyces_fabryi.5ss.tsv\\x94\\x8c\\'pssms_inferred/Meyerozyma_spJA9.5ss.tsv\\x94\\x8c*pssms_inferred/Candida_inconspicua.5ss.tsv\\x94\\x8c4pssms_inferred/Diddensiella_caesifluorescens.5ss.tsv\\x94\\x8c/pssms_inferred/Tortispora_caseinolytica.5ss.tsv\\x94\\x8c%pssms_inferred/Ogataea_minuta.5ss.tsv\\x94\\x8c\\'pssms_inferred/Candida_jaroonii.5ss.tsv\\x94\\x8c0pssms_inferred/Metschnikowia_pulcherrima.5ss.tsv\\x94\\x8c*pssms_inferred/Millerozyma_acaciae.5ss.tsv\\x94\\x8c*pssms_inferred/Tortispora_starmeri.5ss.tsv\\x94\\x8c-pssms_inferred/Hanseniaspora_opuntiae.5ss.tsv\\x94\\x8c)pssms_inferred/Ascoidea_rubescens.5ss.tsv\\x94\\x8c+pssms_inferred/Candida_dubliniensis.5ss.tsv\\x94\\x8c.pssms_inferred/Torulaspora_delbrueckii.5ss.tsv\\x94\\x8c.pssms_inferred/Wickerhamia_fluorescens.5ss.tsv\\x94\\x8c,pssms_inferred/Cyberlindnera_jadinii.5ss.tsv\\x94\\x8c-pssms_inferred/Caenorhabditis_elegans.5ss.tsv\\x94\\x8c+pssms_inferred/Phaffomyces_opuntiae.5ss.tsv\\x94\\x8c)pssms_inferred/Ogataea_polymorpha.5ss.tsv\\x94\\x8c*pssms_inferred/Brettanomyces_nanus.5ss.tsv\\x94\\x8c0pssms_inferred/Wickerhamiella_sorbophila.5ss.tsv\\x94\\x8c0pssms_inferred/Lodderomyces_elongesporus.5ss.tsv\\x94\\x8c&pssms_inferred/Ogataea_zsoltii.5ss.tsv\\x94\\x8c,pssms_inferred/Candida_orthopsilosis.5ss.tsv\\x94\\x8c)pssms_inferred/Diutina_catenulata.5ss.tsv\\x94\\x8c1pssms_inferred/Magnusiomyces_tetraspermus.5ss.tsv\\x94\\x8c/pssms_inferred/Wickerhamomyces_anomalus.5ss.tsv\\x94\\x8c/pssms_inferred/Wickerhamomyces_ciferrii.5ss.tsv\\x94\\x8c*pssms_inferred/Starmerella_apicola.5ss.tsv\\x94\\x8c(pssms_inferred/Pichia_manshurica.5ss.tsv\\x94\\x8c(pssms_inferred/Candida_subhashii.5ss.tsv\\x94\\x8c.pssms_inferred/Saccharomyces_eubayanus.5ss.tsv\\x94\\x8c)pssms_inferred/Kodamaea_laetipori.5ss.tsv\\x94\\x8c+pssms_inferred/Candida_jiufengensis.5ss.tsv\\x94\\x8c/pssms_inferred/Zygosaccharomyces_rouxii.5ss.tsv\\x94\\x8c)pssms_inferred/Saprochaete_ingens.5ss.tsv\\x94\\x8c/pssms_inferred/Spathaspora_passalidarum.5ss.tsv\\x94\\x8c)pssms_inferred/Yamadazyma_scolyti.5ss.tsv\\x94\\x8c,pssms_inferred/Clavispora_lusitaniae.5ss.tsv\\x94\\x8c)pssms_inferred/Lipomyces_starkeyi.5ss.tsv\\x94\\x8c(pssms_inferred/Zygoascus_meyerae.5ss.tsv\\x94\\x8c)pssms_inferred/Pichia_norvegensis.5ss.tsv\\x94\\x8c*pssms_inferred/Ogataea_ramenticola.5ss.tsv\\x94\\x8c)pssms_inferred/Lachancea_meyersii.5ss.tsv\\x94\\x8c+pssms_inferred/Wickerhamomyces_alni.5ss.tsv\\x94\\x8c*pssms_inferred/Candida_oxycetoniae.5ss.tsv\\x94\\x8c/pssms_inferred/Wickerhamiella_cacticola.5ss.tsv\\x94\\x8c/pssms_inferred/Barnettozyma_californica.5ss.tsv\\x94\\x8c.pssms_inferred/Komagataella_kurtzmanii.5ss.tsv\\x94\\x8c-pssms_inferred/Trigonopsis_variabilis.5ss.tsv\\x94\\x8c,pssms_inferred/Deakozyma_indianensis.5ss.tsv\\x94\\x8c*pssms_inferred/Lipomyces_japonicus.5ss.tsv\\x94\\x8c+pssms_inferred/Fusarium_graminearum.5ss.tsv\\x94\\x8c+pssms_inferred/Debaryomyces_maramus.5ss.tsv\\x94\\x8c.pssms_inferred/Ambrosiozyma_philentoma.5ss.tsv\\x94\\x8c/pssms_inferred/Scheffersomyces_stipitis.5ss.tsv\\x94\\x8c)pssms_inferred/Nakazawaea_peltata.5ss.tsv\\x94\\x8c-pssms_inferred/Pneumocystis_jirovecii.5ss.tsv\\x94\\x8c3pssms_inferred/Saturnispora_serradocipensis.5ss.tsv\\x94\\x8c0pssms_inferred/Grigorovia_transvaalensis.5ss.tsv\\x94\\x8c/pssms_inferred/Botryozyma_nematodophila.5ss.tsv\\x94\\x8c/pssms_inferred/Kazachstania_saulgeensis.5ss.tsv\\x94\\x8c.pssms_inferred/Wickerhamomyces_mucosus.5ss.tsv\\x94\\x8c,pssms_inferred/Priceomyces_castillae.5ss.tsv\\x94\\x8c,pssms_inferred/Torulaspora_spCBS2947.5ss.tsv\\x94\\x8c3pssms_inferred/Spencermartinsiella_europaea.5ss.tsv\\x94\\x8c/pssms_inferred/Danielozyma_ontarioensis.5ss.tsv\\x94\\x8c+pssms_inferred/Priceomyces_carsonii.5ss.tsv\\x94\\x8c+pssms_inferred/Meyerozyma_caribbica.5ss.tsv\\x94\\x8c2pssms_inferred/Saccharomycopsis_fermentans.5ss.tsv\\x94\\x8c-pssms_inferred/Blastobotrys_muscicola.5ss.tsv\\x94\\x8c-pssms_inferred/Trichomonascus_ciferii.5ss.tsv\\x94\\x8c+pssms_inferred/Yamadazyma_philogaea.5ss.tsv\\x94\\x8c)pssms_inferred/Nakazawaea_holstii.5ss.tsv\\x94\\x8c#pssms_inferred/Homo_sapiens.5ss.tsv\\x94\\x8c(pssms_inferred/Spathaspora_spJA1.5ss.tsv\\x94\\x8c.pssms_inferred/Hanseniaspora_osmophila.5ss.tsv\\x94\\x8c&pssms_inferred/Ustilago_maydis.5ss.tsv\\x94\\x8c-pssms_inferred/Blastobotrys_mokoenaii.5ss.tsv\\x94\\x8c/pssms_inferred/Candida_duobushaemulonis.5ss.tsv\\x94\\x8c*pssms_inferred/Kurtzmaniella_fragi.5ss.tsv\\x94\\x8c*pssms_inferred/Saturnispora_silvae.5ss.tsv\\x94\\x8c-pssms_inferred/Cyberlindnera_fabianii.5ss.tsv\\x94\\x8c)pssms_inferred/Candida_intermedia.5ss.tsv\\x94\\x8c)pssms_inferred/Priceomyces_medius.5ss.tsv\\x94\\x8c*pssms_inferred/Candida_californica.5ss.tsv\\x94\\x8c-pssms_inferred/Crytococcus_neoformans.5ss.tsv\\x94\\x8c)pssms_inferred/Yueomyces_sinensis.5ss.tsv\\x94\\x8c*pssms_inferred/Kazachstania_exigua.5ss.tsv\\x94\\x8c3pssms_inferred/Zygosaccharomyces_parabailii.5ss.tsv\\x94\\x8c0pssms_inferred/Kluyveromyces_dobzhanskii.5ss.tsv\\x94\\x8c+pssms_inferred/Dekkera_bruxellensis.5ss.tsv\\x94\\x8c%pssms_inferred/Diutina_rugosa.5ss.tsv\\x94\\x8c+pssms_inferred/Candida_hispaniensis.5ss.tsv\\x94\\x8c.pssms_inferred/Naumovozyma_dairenensis.5ss.tsv\\x94\\x8c3pssms_inferred/Sporopachydermia_lactativora.5ss.tsv\\x94\\x8c+pssms_inferred/Starmera_amethionina.5ss.tsv\\x94\\x8c&pssms_inferred/Yarrowia_bubula.5ss.tsv\\x94\\x8c,pssms_inferred/Kazachstania_unispora.5ss.tsv\\x94\\x8c7pssms_inferred/Ambrosiozyma_pseudovanderkliftii.5ss.tsv\\x94\\x8c#pssms_inferred/Ashbya_aceri.5ss.tsv\\x94\\x8c.pssms_inferred/Eremothecium_sinecaudum.5ss.tsv\\x94\\x8c-pssms_inferred/Saturnispora_mendoncae.5ss.tsv\\x94\\x8c\\'pssms_inferred/Candida_albicans.5ss.tsv\\x94\\x8c*pssms_inferred/Pichia_kudriavzevii.5ss.tsv\\x94\\x8c+pssms_inferred/Candida_parapsilosis.5ss.tsv\\x94\\x8c+pssms_inferred/Debaryomyces_hanseni.5ss.tsv\\x94\\x8c&pssms_inferred/Candida_maltosa.5ss.tsv\\x94\\x8c1pssms_inferred/Ambrosiozyma_kashinagicola.5ss.tsv\\x94\\x8c*pssms_inferred/Pichia_occidentalis.5ss.tsv\\x94\\x8c,pssms_inferred/Citeromyces_siamensis.5ss.tsv\\x94\\x8c\\'pssms_inferred/Candida_glabrata.5ss.tsv\\x94\\x8c%pssms_inferred/Pichia_nakasei.5ss.tsv\\x94\\x8c\\'pssms_inferred/Kuraishia_ogatae.5ss.tsv\\x94\\x8c+pssms_inferred/Komagataella_phaffii.5ss.tsv\\x94\\x8c*pssms_inferred/Trigonopsis_vinaria.5ss.tsv\\x94\\x8c+pssms_inferred/Cyberlindnera_mrakii.5ss.tsv\\x94\\x8c)pssms_inferred/Candida_ponderosae.5ss.tsv\\x94\\x8c+pssms_inferred/Aspergillus_nidulans.5ss.tsv\\x94\\x8c&pssms_inferred/Candida_montana.5ss.tsv\\x94\\x8c,pssms_inferred/Kazachstania_africana.5ss.tsv\\x94\\x8c*pssms_inferred/Lachancea_nothofagi.5ss.tsv\\x94\\x8c$pssms_inferred/Candida_theae.5ss.tsv\\x94\\x8c-pssms_inferred/Tetrapisispora_phaffii.5ss.tsv\\x94\\x8c(pssms_inferred/Starmera_quercuum.5ss.tsv\\x94\\x8c1pssms_inferred/Sugiyamaella_lignohabitans.5ss.tsv\\x94\\x8c)pssms_inferred/Teunomyces_kruisii.5ss.tsv\\x94\\x8c/pssms_inferred/Kurtzmaniella_cleridarum.5ss.tsv\\x94\\x8c.pssms_inferred/Magnusiomyces_capitatus.5ss.tsv\\x94\\x8c/pssms_inferred/Saccharomyces_cerevisiae.5ss.tsv\\x94\\x8c(pssms_inferred/Peterozyma_xylosa.5ss.tsv\\x94\\x8c+pssms_inferred/Saturnispora_dispora.5ss.tsv\\x94\\x8c)pssms_inferred/Ogataea_pilisensis.5ss.tsv\\x94\\x8c*pssms_inferred/Kuraishia_capsulata.5ss.tsv\\x94\\x8c7pssms_inferred/Blastobotrys_raffinosifermentans.5ss.tsv\\x94\\x8c+pssms_inferred/Eremothecium_gosypii.5ss.tsv\\x94\\x8c0pssms_inferred/Scheffersomyces_spartinae.5ss.tsv\\x94\\x8c,pssms_inferred/Kockiozyma_suomiensis.5ss.tsv\\x94\\x8c+pssms_inferred/Cephaloascus_albidus.5ss.tsv\\x94\\x8c(pssms_inferred/Neurospora_crassa.5ss.tsv\\x94\\x8c,pssms_inferred/Lachancea_quebecensis.5ss.tsv\\x94\\x8c.pssms_inferred/Citeromyces_matritensis.5ss.tsv\\x94\\x8c&pssms_inferred/Ogataea_kodamae.5ss.tsv\\x94\\x8c0pssms_inferred/Vanderwaltozyma_polyspora.5ss.tsv\\x94\\x8c-pssms_inferred/Barnettozyma_pratensis.5ss.tsv\\x94\\x8c-pssms_inferred/Candida_ascalaphidarum.5ss.tsv\\x94\\x8c/pssms_inferred/Eremothecium_cymbalariae.5ss.tsv\\x94\\x8c.pssms_inferred/Saccharomyces_paradoxus.5ss.tsv\\x94\\x8c-pssms_inferred/Blastobotrys_aristatus.5ss.tsv\\x94\\x8c\\'pssms_inferred/Botrytis_cinerea.5ss.tsv\\x94\\x8c,pssms_inferred/Cephaloascus_fragrans.5ss.tsv\\x94\\x8c(pssms_inferred/Candida_chilensis.5ss.tsv\\x94\\x8c/pssms_inferred/Saccharomycopsis_fodiens.5ss.tsv\\x94\\x8c.pssms_inferred/Wickerhamomyces_pijperi.5ss.tsv\\x94\\x8c*pssms_inferred/Ogataea_methylovora.5ss.tsv\\x94\\x8c,pssms_inferred/Naumovozyma_castellii.5ss.tsv\\x94\\x8c0pssms_inferred/Sporopachydermia_quercuum.5ss.tsv\\x94\\x8c\\'pssms_inferred/Candida_boidinii.5ss.tsv\\x94\\x8c\\'pssms_inferred/Ogataea_henricii.5ss.tsv\\x94\\x8c+pssms_inferred/Ambrosiozyma_maleeae.5ss.tsv\\x94\\x8c*pssms_inferred/Kluveromyces_lactis.5ss.tsv\\x94\\x8c-pssms_inferred/Saturnispora_zaruensis.5ss.tsv\\x94\\x8c-pssms_inferred/Ogataea_parapolymorpha.5ss.tsv\\x94\\x8c0pssms_inferred/Candida_arabinofermentans.5ss.tsv\\x94\\x8c)pssms_inferred/Tortispora_ganteri.5ss.tsv\\x94\\x8c*pssms_inferred/Torulaspora_globosa.5ss.tsv\\x94\\x8c)pssms_inferred/Ogataea_naganishii.5ss.tsv\\x94\\x8c.pssms_inferred/Saccharomyces_boulardii.5ss.tsv\\x94\\x8c/pssms_inferred/Zygosaccharomyces_bailii.5ss.tsv\\x94\\x8c/pssms_inferred/Lachancea_thermotolerans.5ss.tsv\\x94\\x8c-pssms_inferred/Suhomyces_tanzawaensis.5ss.tsv\\x94\\x8c+pssms_inferred/Zygoascus_ofunaensis.5ss.tsv\\x94\\x8c-pssms_inferred/Pachysolen_tannophilus.5ss.tsv\\x94\\x8c/pssms_inferred/Saccharomycopsis_malanga.5ss.tsv\\x94\\x8c.pssms_inferred/Suhomyces_canberraensis.5ss.tsv\\x94\\x8c-pssms_inferred/Kazachstania_barnettii.5ss.tsv\\x94\\x8c0pssms_inferred/Nadsonia_starkeyihenricii.5ss.tsv\\x94\\x8c.pssms_inferred/Sugiyamaella_xylanicola.5ss.tsv\\x94\\x8c(pssms_inferred/Hyphopichia_gotoi.5ss.tsv\\x94\\x8c.pssms_inferred/Kluyveromyces_marxianus.5ss.tsv\\x94\\x8c(pssms_inferred/Ascoidea_asiatica.5ss.tsv\\x94\\x8c,pssms_inferred/Starmerella_bombicola.5ss.tsv\\x94\\x8c/pssms_inferred/Ogataea_trehaloabstinens.5ss.tsv\\x94\\x8c+pssms_inferred/Lachancea_fermentati.5ss.tsv\\x94\\x8c*pssms_inferred/Candida_railenensis.5ss.tsv\\x94\\x8c.pssms_inferred/Lachancea_lanzerotensis.5ss.tsv\\x94\\x8c0pssms_inferred/Meyerozyma_guilliermondii.5ss.tsv\\x94\\x8c+pssms_inferred/Hanseniaspora_uvarum.5ss.tsv\\x94\\x8c.pssms_inferred/Middelhovenomyces_tepae.5ss.tsv\\x94\\x8c(pssms_inferred/Candida_schatavii.5ss.tsv\\x94\\x8c-pssms_inferred/Meyerozyma_athensensis.5ss.tsv\\x94\\x8c0pssms_inferred/Wickerhamiella_domercqiae.5ss.tsv\\x94\\x8c-pssms_inferred/Lipomyces_doorenjongii.5ss.tsv\\x94\\x8c$pssms_inferred/Pichia_exigua.5ss.tsv\\x94\\x8c/pssms_inferred/Scheffersomyces_lignosus.5ss.tsv\\x94\\x8c/pssms_inferred/Zygosaccharomyces_mellis.5ss.tsv\\x94\\x8c/pssms_inferred/Martiniozyma_abietophila.5ss.tsv\\x94\\x8c*pssms_inferred/Lachancea_mirantina.5ss.tsv\\x94\\x8c(pssms_inferred/Candida_haemuloni.5ss.tsv\\x94\\x8c.pssms_inferred/Babjeviella_inositovora.5ss.tsv\\x94\\x8c,pssms_inferred/Saprochaete_fungicola.5ss.tsv\\x94\\x8c*pssms_inferred/Lachancea_dasiensis.5ss.tsv\\x94\\x8c.pssms_inferred/Hyphopichia_homilentoma.5ss.tsv\\x94\\x8c/pssms_inferred/Candida_pseudohaemulonii.5ss.tsv\\x94\\x8c0pssms_inferred/Grigorovia_yakushimaensis.5ss.tsv\\x94\\x8c5pssms_inferred/Groenewaldozyma_salmanticensis.5ss.tsv\\x94\\x8c,pssms_inferred/Komagataella_pastoris.5ss.tsv\\x94\\x8c\\x1corthogroups/species_tree.txt\\x94\\x8c+key_prot_ogs/orthogroups/orthogroups_n0.tsv\\x94\\x8c&../../annotations/splicing_factors.tsv\\x94e}\\x94(\\x8c\\x06_names\\x94}\\x94(\\x8c\\x05pssms\\x94K\\x00K\\xf0\\x86\\x94\\x8c\\x04tree\\x94K\\xf0N\\x86\\x94\\x8c\\x07key_ogs\\x94K\\xf1N\\x86\\x94\\x8c\\x0ckey_prot_ids\\x94K\\xf2N\\x86\\x94u\\x8c\\x12_allowed_overrides\\x94]\\x94(\\x8c\\x05index\\x94\\x8c\\x04sort\\x94ej\\n\\x01\\x00\\x00\\x8c\\tfunctools\\x94\\x8c\\x07partial\\x94\\x93\\x94h\\x06\\x8c\\x19Namedlist._used_attribute\\x94\\x93\\x94\\x85\\x94R\\x94(j\\x10\\x01\\x00\\x00)}\\x94\\x8c\\x05_name\\x94j\\n\\x01\\x00\\x00sNt\\x94bj\\x0b\\x01\\x00\\x00j\\x0e\\x01\\x00\\x00j\\x10\\x01\\x00\\x00\\x85\\x94R\\x94(j\\x10\\x01\\x00\\x00)}\\x94j\\x14\\x01\\x00\\x00j\\x0b\\x01\\x00\\x00sNt\\x94bj\\x00\\x01\\x00\\x00h\\x06\\x8c\\tNamedlist\\x94\\x93\\x94)\\x81\\x94(h\\nh\\x0bh\\x0ch\\rh\\x0eh\\x0fh\\x10h\\x11h\\x12h\\x13h\\x14h\\x15h\\x16h\\x17h\\x18h\\x19h\\x1ah\\x1bh\\x1ch\\x1dh\\x1eh\\x1fh h!h\"h#h$h%h&h\\'h(h)h*h+h,h-h.h/h0h1h2h3h4h5h6h7h8h9h:h;h<h=h>h?h@hAhBhChDhEhFhGhHhIhJhKhLhMhNhOhPhQhRhShThUhVhWhXhYhZh[h\\\\h]h^h_h`hahbhchdhehfhghhhihjhkhlhmhnhohphqhrhshthuhvhwhxhyhzh{h|h}h~h\\x7fh\\x80h\\x81h\\x82h\\x83h\\x84h\\x85h\\x86h\\x87h\\x88h\\x89h\\x8ah\\x8bh\\x8ch\\x8dh\\x8eh\\x8fh\\x90h\\x91h\\x92h\\x93h\\x94h\\x95h\\x96h\\x97h\\x98h\\x99h\\x9ah\\x9bh\\x9ch\\x9dh\\x9eh\\x9fh\\xa0h\\xa1h\\xa2h\\xa3h\\xa4h\\xa5h\\xa6h\\xa7h\\xa8h\\xa9h\\xaah\\xabh\\xach\\xadh\\xaeh\\xafh\\xb0h\\xb1h\\xb2h\\xb3h\\xb4h\\xb5h\\xb6h\\xb7h\\xb8h\\xb9h\\xbah\\xbbh\\xbch\\xbdh\\xbeh\\xbfh\\xc0h\\xc1h\\xc2h\\xc3h\\xc4h\\xc5h\\xc6h\\xc7h\\xc8h\\xc9h\\xcah\\xcbh\\xcch\\xcdh\\xceh\\xcfh\\xd0h\\xd1h\\xd2h\\xd3h\\xd4h\\xd5h\\xd6h\\xd7h\\xd8h\\xd9h\\xdah\\xdbh\\xdch\\xddh\\xdeh\\xdfh\\xe0h\\xe1h\\xe2h\\xe3h\\xe4h\\xe5h\\xe6h\\xe7h\\xe8h\\xe9h\\xeah\\xebh\\xech\\xedh\\xeeh\\xefh\\xf0h\\xf1h\\xf2h\\xf3h\\xf4h\\xf5h\\xf6h\\xf7h\\xf8h\\xf9e}\\x94(h\\xfe}\\x94j\\x08\\x01\\x00\\x00]\\x94(j\\n\\x01\\x00\\x00j\\x0b\\x01\\x00\\x00ej\\n\\x01\\x00\\x00j\\x0e\\x01\\x00\\x00j\\x10\\x01\\x00\\x00\\x85\\x94R\\x94(j\\x10\\x01\\x00\\x00)}\\x94j\\x14\\x01\\x00\\x00j\\n\\x01\\x00\\x00sNt\\x94bj\\x0b\\x01\\x00\\x00j\\x0e\\x01\\x00\\x00j\\x10\\x01\\x00\\x00\\x85\\x94R\\x94(j\\x10\\x01\\x00\\x00)}\\x94j\\x14\\x01\\x00\\x00j\\x0b\\x01\\x00\\x00sNt\\x94bubj\\x02\\x01\\x00\\x00h\\xfaj\\x04\\x01\\x00\\x00h\\xfbj\\x06\\x01\\x00\\x00h\\xfcub\\x8c\\x06output\\x94h\\x06\\x8c\\x0bOutputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8c figures/5ss.p4AtoU.phenogram.svg\\x94\\x8c-figures/5ss.p4AtoU.tree_with_stacked_bars.svg\\x94\\x8c2figures/5ss.p4AtoU.tree_with_stacked_bars_cbar.svg\\x94\\x8c$figures/5ss.p4AtoU.circular_tree.svg\\x94e}\\x94(h\\xfe}\\x94(\\x8c\\tphenogram\\x94K\\x00N\\x86\\x94\\x8c\\x0ecollapsed_tree\\x94K\\x01N\\x86\\x94\\x8c\\x13collapsed_tree_cbar\\x94K\\x02N\\x86\\x94\\x8c\\x12full_circular_tree\\x94K\\x03N\\x86\\x94uj\\x08\\x01\\x00\\x00]\\x94(j\\n\\x01\\x00\\x00j\\x0b\\x01\\x00\\x00ej\\n\\x01\\x00\\x00j\\x0e\\x01\\x00\\x00j\\x10\\x01\\x00\\x00\\x85\\x94R\\x94(j\\x10\\x01\\x00\\x00)}\\x94j\\x14\\x01\\x00\\x00j\\n\\x01\\x00\\x00sNt\\x94bj\\x0b\\x01\\x00\\x00j\\x0e\\x01\\x00\\x00j\\x10\\x01\\x00\\x00\\x85\\x94R\\x94(j\\x10\\x01\\x00\\x00)}\\x94j\\x14\\x01\\x00\\x00j\\x0b\\x01\\x00\\x00sNt\\x94bj2\\x01\\x00\\x00j,\\x01\\x00\\x00j4\\x01\\x00\\x00j-\\x01\\x00\\x00j6\\x01\\x00\\x00j.\\x01\\x00\\x00j8\\x01\\x00\\x00j/\\x01\\x00\\x00ub\\x8c\\x06params\\x94h\\x06\\x8c\\x06Params\\x94\\x93\\x94)\\x81\\x94}\\x94(h\\xfe}\\x94j\\x08\\x01\\x00\\x00]\\x94(j\\n\\x01\\x00\\x00j\\x0b\\x01\\x00\\x00ej\\n\\x01\\x00\\x00j\\x0e\\x01\\x00\\x00j\\x10\\x01\\x00\\x00\\x85\\x94R\\x94(j\\x10\\x01\\x00\\x00)}\\x94j\\x14\\x01\\x00\\x00j\\n\\x01\\x00\\x00sNt\\x94bj\\x0b\\x01\\x00\\x00j\\x0e\\x01\\x00\\x00j\\x10\\x01\\x00\\x00\\x85\\x94R\\x94(j\\x10\\x01\\x00\\x00)}\\x94j\\x14\\x01\\x00\\x00j\\x0b\\x01\\x00\\x00sNt\\x94bub\\x8c\\twildcards\\x94h\\x06\\x8c\\tWildcards\\x94\\x93\\x94)\\x81\\x94(\\x8c\\x035ss\\x94\\x8c\\x06p4AtoU\\x94e}\\x94(h\\xfe}\\x94(\\x8c\\x0bsignal_type\\x94K\\x00N\\x86\\x94\\x8c\\tphenotype\\x94K\\x01N\\x86\\x94uj\\x08\\x01\\x00\\x00]\\x94(j\\n\\x01\\x00\\x00j\\x0b\\x01\\x00\\x00ej\\n\\x01\\x00\\x00j\\x0e\\x01\\x00\\x00j\\x10\\x01\\x00\\x00\\x85\\x94R\\x94(j\\x10\\x01\\x00\\x00)}\\x94j\\x14\\x01\\x00\\x00j\\n\\x01\\x00\\x00sNt\\x94bj\\x0b\\x01\\x00\\x00j\\x0e\\x01\\x00\\x00j\\x10\\x01\\x00\\x00\\x85\\x94R\\x94(j\\x10\\x01\\x00\\x00)}\\x94j\\x14\\x01\\x00\\x00j\\x0b\\x01\\x00\\x00sNt\\x94bjZ\\x01\\x00\\x00jV\\x01\\x00\\x00j\\\\\\x01\\x00\\x00jW\\x01\\x00\\x00ub\\x8c\\x07threads\\x94K\\x01\\x8c\\tresources\\x94h\\x06\\x8c\\tResources\\x94\\x93\\x94)\\x81\\x94(K\\x01K\\x01\\x8c\\x14/tmp/1175317.1.all.q\\x94e}\\x94(h\\xfe}\\x94(\\x8c\\x06_cores\\x94K\\x00N\\x86\\x94\\x8c\\x06_nodes\\x94K\\x01N\\x86\\x94\\x8c\\x06tmpdir\\x94K\\x02N\\x86\\x94uj\\x08\\x01\\x00\\x00]\\x94(j\\n\\x01\\x00\\x00j\\x0b\\x01\\x00\\x00ej\\n\\x01\\x00\\x00j\\x0e\\x01\\x00\\x00j\\x10\\x01\\x00\\x00\\x85\\x94R\\x94(j\\x10\\x01\\x00\\x00)}\\x94j\\x14\\x01\\x00\\x00j\\n\\x01\\x00\\x00sNt\\x94bj\\x0b\\x01\\x00\\x00j\\x0e\\x01\\x00\\x00j\\x10\\x01\\x00\\x00\\x85\\x94R\\x94(j\\x10\\x01\\x00\\x00)}\\x94j\\x14\\x01\\x00\\x00j\\x0b\\x01\\x00\\x00sNt\\x94bjo\\x01\\x00\\x00K\\x01jq\\x01\\x00\\x00K\\x01js\\x01\\x00\\x00jl\\x01\\x00\\x00ub\\x8c\\x03log\\x94h\\x06\\x8c\\x03Log\\x94\\x93\\x94)\\x81\\x94\\x8c5notebook_processed/draw_phenogram.5ss.p4AtoU.py.ipynb\\x94a}\\x94(h\\xfe}\\x94\\x8c\\x08notebook\\x94K\\x00N\\x86\\x94sj\\x08\\x01\\x00\\x00]\\x94(j\\n\\x01\\x00\\x00j\\x0b\\x01\\x00\\x00ej\\n\\x01\\x00\\x00j\\x0e\\x01\\x00\\x00j\\x10\\x01\\x00\\x00\\x85\\x94R\\x94(j\\x10\\x01\\x00\\x00)}\\x94j\\x14\\x01\\x00\\x00j\\n\\x01\\x00\\x00sNt\\x94bj\\x0b\\x01\\x00\\x00j\\x0e\\x01\\x00\\x00j\\x10\\x01\\x00\\x00\\x85\\x94R\\x94(j\\x10\\x01\\x00\\x00)}\\x94j\\x14\\x01\\x00\\x00j\\x0b\\x01\\x00\\x00sNt\\x94bj\\x85\\x01\\x00\\x00j\\x82\\x01\\x00\\x00ub\\x8c\\x06config\\x94}\\x94(\\x8c\\x05clade\\x94\\x8c\\nascomycota\\x94\\x8c\\x11key_prot_table_fn\\x94\\x8c&../../annotations/splicing_factors.tsv\\x94\\x8c\\x13reference_orthologs\\x94}\\x94\\x8c\\x19Schizosaccharomyces_pombe\\x94}\\x94\\x8c\\x07METTL16\\x94\\x8c\\x0bNP_594614.1\\x94ss\\x8c\\x05busco\\x94}\\x94(\\x8c\\x15default_busco_lineage\\x94\\x8c\\nascomycota\\x94\\x8c\\x0ebusco_lineages\\x94}\\x94(\\x8c\\x0cHomo_sapiens\\x94\\x8c\\x08primates\\x94\\x8c\\x16Caenorhabditis_elegans\\x94\\x8c\\x08nematoda\\x94\\x8c\\x17Cryptococcus_neoformans\\x94\\x8c\\x0ftremellomycetes\\x94\\x8c\\x0fUstilago_maydis\\x94\\x8c\\rbasidiomycota\\x94uu\\x8c\\x0bfunannotate\\x94}\\x94(\\x8c\\x13trusted_annotations\\x94]\\x94(\\x8c\\x0cHomo_sapiens\\x94\\x8c\\x16Caenorhabditis_elegans\\x94\\x8c\\x16Crytococcus_neoformans\\x94\\x8c\\x0fUstilago_maydis\\x94\\x8c\\x19Schizosaccharomyces_pombe\\x94\\x8c\\x16Pneumocystis_jirovecii\\x94\\x8c\\x14Aspergillus_nidulans\\x94\\x8c\\x10Botrytis_cinerea\\x94\\x8c\\x11Neurospora_crassa\\x94\\x8c\\x14Fusarium_graminearum\\x94\\x8c\\x13Yarrowia_lipolytica\\x94\\x8c\\x10Candida_albicans\\x94\\x8c\\x18Saccharomyces_cerevisiae\\x94e\\x8c\\x06ploidy\\x94}\\x94(\\x8c\\x1cZygosaccharomyces_parabailii\\x94K\\x02\\x8c\\x19Saccharomyces_pastorianus\\x94K\\x02\\x8c\\x14Millerozyma_farinosa\\x94K\\x02u\\x8c\\x0cseed_species\\x94}\\x94(\\x8c\\x13yarrowia_lipolytica\\x94]\\x94(\\x8c\\x12Lipomyces_starkeyi\\x94\\x8c\\x13Lipomyces_japonicus\\x94\\x8c\\x14Lipomyces_oligophaga\\x94\\x8c\\x16Lipomyces_doorenjongii\\x94\\x8c\\x15Kockiozyma_suomiensis\\x94\\x8c\\x16Trigonopsis_variabilis\\x94\\x8c\\x13Trigonopsis_vinaria\\x94\\x8c\\x18Botryozyma_nematodophila\\x94\\x8c\\x13Tortispora_starmeri\\x94\\x8c\\x18Tortispora_caseinolytica\\x94\\x8c\\x12Tortispora_ganteri\\x94\\x8c\\x14Candida_hispaniensis\\x94\\x8c\\x0fYarrowia_bubula\\x94\\x8c\\x15Deakozyma_indianensis\\x94\\x8c\\x19Wickerhamiella_sorbophila\\x94\\x8c\\x18Wickerhamiella_cacticola\\x94\\x8c\\x19Wickerhamiella_domercqiae\\x94\\x8c\\x19Wickerhamiella_versatilis\\x94\\x8c\\x13Starmerella_apicola\\x94\\x8c\\x15Starmerella_bombicola\\x94\\x8c\\x16Trichomonascus_ciferii\\x94\\x8c\\x12Saprochaete_ingens\\x94\\x8c\\x15Saprochaete_fungicola\\x94\\x8c\\x13Nadsonia_fulvescens\\x94\\x8c\\x19Nadsonia_starkeyihenricii\\x94\\x8c\\x17Middelhovenomyces_tepae\\x94\\x8c\\x17Magnusiomyces_capitatus\\x94\\x8c\\x1aMagnusiomyces_tetraspermus\\x94\\x8c\\x12Dipodascus_albidus\\x94\\x8c\\x1eGroenewaldozyma_salmanticensis\\x94\\x8c\\x14Zygoascus_ofunaensis\\x94\\x8c\\x11Zygoascus_meyerae\\x94\\x8c\\x1dDiddensiella_caesifluorescens\\x94\\x8c\\x1cSpencermartinsiella_europaea\\x94\\x8c\\x16Blastobotrys_muscicola\\x94\\x8c\\x16Blastobotrys_mokoenaii\\x94\\x8c\\x16Blastobotrys_aristatus\\x94\\x8c Blastobotrys_raffinosifermentans\\x94\\x8c\\x1aSugiyamaella_lignohabitans\\x94\\x8c\\x17Sugiyamaella_xylanicola\\x94e\\x8c\\x10candida_albicans\\x94]\\x94(\\x8c\\x16Alloascoidea_hylecoeti\\x94\\x8c\\x19Sporopachydermia_quercuum\\x94\\x8c\\x1cSporopachydermia_lactativora\\x94\\x8c\\x16Pachysolen_tannophilus\\x94\\x8c\\x11Peterozyma_xylosa\\x94\\x8c\\x12Nakazawaea_peltata\\x94\\x8c\\x12Nakazawaea_holstii\\x94\\x8c\\x13Kuraishia_capsulata\\x94\\x8c\\x10Kuraishia_ogatae\\x94\\x8c\\x10Candida_boidinii\\x94\\x8c\\x19Candida_arabinofermentans\\x94\\x8c\\x16Ogataea_parapolymorpha\\x94\\x8c\\x12Ogataea_polymorpha\\x94\\x8c\\x13Ogataea_methylovora\\x94\\x8c\\x18Ogataea_trehaloabstinens\\x94\\x8c\\x0fOgataea_kodamae\\x94\\x8c\\x0eOgataea_minuta\\x94\\x8c\\x10Ogataea_henricii\\x94\\x8c\\x0fOgataea_zsoltii\\x94\\x8c\\x12Ogataea_pilisensis\\x94\\x8c\\x12Ogataea_naganishii\\x94\\x8c\\x13Ogataea_ramenticola\\x94\\x8c\\x14Ambrosiozyma_maleeae\\x94\\x8c Ambrosiozyma_pseudovanderkliftii\\x94\\x8c\\x17Ambrosiozyma_philentoma\\x94\\x8c\\x1aAmbrosiozyma_kashinagicola\\x94\\x8c\\x14Dekkera_bruxellensis\\x94\\x8c\\x13Brettanomyces_nanus\\x94\\x8c\\x1aBrettanomyces_naardenensis\\x94\\x8c\\x18Kregervanrija_delftensis\\x94\\x8c\\x15Kregervanrija_fluxuum\\x94\\x8c\\x18Martiniozyma_abietophila\\x94\\x8c\\x13Saturnispora_silvae\\x94\\x8c\\x1cSaturnispora_serradocipensis\\x94\\x8c\\x16Saturnispora_zaruensis\\x94\\x8c\\x16Saturnispora_mendoncae\\x94\\x8c\\x14Saturnispora_dispora\\x94\\x8c\\x16Pichia_membranifaciens\\x94\\x8c\\x13Candida_californica\\x94\\x8c\\x13Pichia_kudriavzevii\\x94\\x8c\\x10Pichia_terricola\\x94\\x8c\\x12Pichia_norvegensis\\x94\\x8c\\x0ePichia_nakasei\\x94\\x8c\\rPichia_exigua\\x94\\x8c\\x11Pichia_manshurica\\x94\\x8c\\x13Pichia_occidentalis\\x94\\x8c\\x13Candida_inconspicua\\x94\\x8c\\x17Komagataella_kurtzmanii\\x94\\x8c\\x14Komagataella_phaffii\\x94\\x8c\\x15Komagataella_pastoris\\x94\\x8c\\x15Citeromyces_siamensis\\x94\\x8c\\x17Citeromyces_matritensis\\x94\\x8c\\x17Babjeviella_inositovora\\x94\\x8c\\x15Cephaloascus_fragrans\\x94\\x8c\\x14Cephaloascus_albidus\\x94\\x8c\\x0eDiutina_rugosa\\x94\\x8c\\x12Diutina_catenulata\\x94\\x8c\\x11Candida_schatavii\\x94\\x8c\\x18Kurtzmaniella_cleridarum\\x94\\x8c\\x13Kurtzmaniella_fragi\\x94\\x8c\\x16Candida_ascalaphidarum\\x94\\x8c\\x10Candida_gorgasii\\x94\\x8c\\x19Scheffersomyces_spartinae\\x94\\x8c\\x10Candida_jaroonii\\x94\\x8c\\x11Yamadazyma_tenuis\\x94\\x8c\\x14Yamadazyma_philogaea\\x94\\x8c\\x12Yamadazyma_scolyti\\x94\\x8c\\x13Candida_railenensis\\x94\\x8c\\x13Millerozyma_acaciae\\x94\\x8c\\x14Millerozyma_farinosa\\x94\\x8c\\x13Debaryomyces_fabryi\\x94\\x8c\\x14Debaryomyces_hanseni\\x94\\x8c\\x14Debaryomyces_maramus\\x94\\x8c\\x14Priceomyces_carsonii\\x94\\x8c\\x12Priceomyces_medius\\x94\\x8c\\x15Priceomyces_castillae\\x94\\x8c\\x17Wickerhamia_fluorescens\\x94\\x8c\\x15Teunomyces_gatunensis\\x94\\x8c\\x12Teunomyces_kruisii\\x94\\x8c\\x13Suhomyces_pyralidae\\x94\\x8c\\x17Suhomyces_canberraensis\\x94\\x8c\\x12Kodamaea_laetipori\\x94\\x8c\\x18Scheffersomyces_lignosus\\x94\\x8c\\x18Danielozyma_ontarioensis\\x94\\x8c\\x14Hyphopichia_burtonii\\x94\\x8c\\x17Hyphopichia_homilentoma\\x94\\x8c\\x11Hyphopichia_gotoi\\x94\\x8c\\x19Meyerozyma_guilliermondii\\x94\\x8c\\x10Meyerozyma_spJA9\\x94\\x8c\\x14Meyerozyma_caribbica\\x94\\x8c\\x16Meyerozyma_athensensis\\x94\\x8c\\rCandida_auris\\x94\\x8c\\x11Candida_haemuloni\\x94\\x8c\\x18Candida_pseudohaemulonii\\x94\\x8c\\x18Candida_duobushaemulonis\\x94\\x8c\\x18Metschnikowia_spJCM33374\\x94\\x8c\\x19Metschnikowia_bicuspidata\\x94\\x8c\\x19Metschnikowia_pulcherrima\\x94\\x8c\\x12Candida_intermedia\\x94\\x8c\\x15Clavispora_lusitaniae\\x94\\x8c\\x16Suhomyces_tanzawaensis\\x94\\x8c\\x18Scheffersomyces_stipitis\\x94\\x8c\\x11Candida_subhashii\\x94\\x8c\\x18Spathaspora_passalidarum\\x94\\x8c\\x11Spathaspora_spJA1\\x94\\x8c\\x14Candida_dubliniensis\\x94\\x8c\\x11Candida_tropcalis\\x94\\x8c\\x0fCandida_maltosa\\x94\\x8c\\x13Candida_oxycetoniae\\x94\\x8c\\x19Lodderomyces_elongesporus\\x94\\x8c\\x14Candida_jiufengensis\\x94\\x8c\\rCandida_theae\\x94\\x8c\\x14Candida_metapsilosis\\x94\\x8c\\x14Candida_parapsilosis\\x94\\x8c\\x15Candida_orthopsilosis\\x94\\x8c\\x11Candida_chilensis\\x94e\\x8c\\x1esaccharomyces_cerevisiae_S288C\\x94]\\x94(\\x8c\\x1bSaccharomycopsis_fibuligera\\x94\\x8c\\x18Saccharomycopsis_malanga\\x94\\x8c\\x1bSaccharomycopsis_fermentans\\x94\\x8c\\x18Saccharomycopsis_fodiens\\x94\\x8c\\x12Ascoidea_rubescens\\x94\\x8c\\x11Ascoidea_asiatica\\x94\\x8c\\x12Candida_ponderosae\\x94\\x8c\\x11Starmera_quercuum\\x94\\x8c\\x14Starmera_amethionina\\x94\\x8c\\x17Wickerhamomyces_pijperi\\x94\\x8c\\x17Wickerhamomyces_mucosus\\x94\\x8c\\x18Wickerhamomyces_ciferrii\\x94\\x8c\\x18Wickerhamomyces_anomalus\\x94\\x8c\\x14Wickerhamomyces_alni\\x94\\x8c\\x15Wickerhamomyces_bovis\\x94\\x8c\\x15Cyberlindnera_jadinii\\x94\\x8c\\x16Cyberlindnera_fabianii\\x94\\x8c\\x14Cyberlindnera_mrakii\\x94\\x8c\\x18Barnettozyma_californica\\x94\\x8c\\x16Barnettozyma_salicaria\\x94\\x8c\\x16Barnettozyma_pratensis\\x94\\x8c\\x0fCandida_montana\\x94\\x8c\\x14Phaffomyces_opuntiae\\x94\\x8c\\x17Saccharomycodes_ludwigi\\x94\\x8c\\x17Hanseniaspora_osmophila\\x94\\x8c\\x1cHanseniaspora_guilliermondii\\x94\\x8c\\x16Hanseniaspora_opuntiae\\x94\\x8c\\x14Hanseniaspora_uvarum\\x94\\x8c\\x16Tetrapisispora_blattae\\x94\\x8c\\x19Vanderwaltozyma_polyspora\\x94\\x8c\\x16Tetrapisispora_phaffii\\x94\\x8c\\x18Zygosaccharomyces_mellis\\x94\\x8c\\x18Zygosaccharomyces_rouxii\\x94\\x8c\\x18Zygosaccharomyces_bailii\\x94\\x8c\\x1cZygosaccharomyces_parabailii\\x94\\x8c\\x16Zygotorulaspora_mrakii\\x94\\x8c\\x13Torulaspora_globosa\\x94\\x8c\\x15Torulaspora_spCBS2947\\x94\\x8c\\x17Torulaspora_delbrueckii\\x94\\x8c\\x10Candida_glabrata\\x94\\x8c\\x17Saccharomyces_paradoxus\\x94\\x8c\\x19Saccharomyces_pastorianus\\x94\\x8c\\x17Saccharomyces_eubayanus\\x94\\x8c\\x17Saccharomyces_boulardii\\x94\\x8c\\x15Naumovozyma_castellii\\x94\\x8c\\x17Naumovozyma_dairenensis\\x94\\x8c\\x1aNakaseomyces_bacillisporus\\x94\\x8c\\x15Kazachstania_africana\\x94\\x8c\\x13Kazachstania_exigua\\x94\\x8c\\x18Kazachstania_saulgeensis\\x94\\x8c\\x16Kazachstania_barnettii\\x94\\x8c\\x17Kazachstania_naganishii\\x94\\x8c\\x15Kazachstania_unispora\\x94\\x8c\\x17Eremothecium_sinecaudum\\x94\\x8c\\x18Eremothecium_cymbalariae\\x94\\x8c\\x14Eremothecium_gosypii\\x94\\x8c\\x0cAshbya_aceri\\x94\\x8c\\x17Kluyveromyces_marxianus\\x94\\x8c\\x13Kluveromyces_lactis\\x94\\x8c\\x19Kluyveromyces_dobzhanskii\\x94\\x8c\\x13Lachancea_mirantina\\x94\\x8c\\x14Lachancea_fermentati\\x94\\x8c\\x15Lachancea_quebecensis\\x94\\x8c\\x18Lachancea_thermotolerans\\x94\\x8c\\x13Lachancea_nothofagi\\x94\\x8c\\x13Lachancea_dasiensis\\x94\\x8c\\x17Lachancea_lanzerotensis\\x94\\x8c\\x12Lachancea_meyersii\\x94\\x8c\\x12Yueomyces_sinensis\\x94\\x8c\\x19Grigorovia_yakushimaensis\\x94\\x8c\\x19Grigorovia_transvaalensis\\x94euu\\x8c\\x0cknown_clades\\x94}\\x94(\\x8c\\tOutgroups\\x94]\\x94(\\x8c\\x0cHomo_sapiens\\x94\\x8c\\x16Caenorhabditis_elegans\\x94\\x8c\\x16Crytococcus_neoformans\\x94\\x8c\\x0fUstilago_maydis\\x94\\x8c\\x19Schizosaccharomyces_pombe\\x94\\x8c\\x16Pneumocystis_jirovecii\\x94\\x8c\\x14Aspergillus_nidulans\\x94\\x8c\\x10Botrytis_cinerea\\x94\\x8c\\x11Neurospora_crassa\\x94\\x8c\\x14Fusarium_graminearum\\x94e\\x8c\\x10Trigonopsidaceae\\x94]\\x94(\\x8c\\x16Trigonopsis_variabilis\\x94\\x8c\\x13Trigonopsis_vinaria\\x94\\x8c\\x18Botryozyma_nematodophila\\x94\\x8c\\x13Tortispora_starmeri\\x94\\x8c\\x18Tortispora_caseinolytica\\x94\\x8c\\x12Tortispora_ganteri\\x94e\\x8c\\x0eLipomycetaceae\\x94]\\x94(\\x8c\\x12Lipomyces_starkeyi\\x94\\x8c\\x13Lipomyces_japonicus\\x94\\x8c\\x14Lipomyces_oligophaga\\x94\\x8c\\x16Lipomyces_doorenjongii\\x94\\x8c\\x15Kockiozyma_suomiensis\\x94e\\x8c\\rDipodascaceae\\x94]\\x94(\\x8c\\x14Candida_hispaniensis\\x94\\x8c\\x0fYarrowia_bubula\\x94\\x8c\\x13Yarrowia_lipolytica\\x94\\x8c\\x15Deakozyma_indianensis\\x94\\x8c\\x19Wickerhamiella_sorbophila\\x94\\x8c\\x18Wickerhamiella_cacticola\\x94\\x8c\\x19Wickerhamiella_domercqiae\\x94\\x8c\\x19Wickerhamiella_versatilis\\x94\\x8c\\x13Starmerella_apicola\\x94\\x8c\\x15Starmerella_bombicola\\x94\\x8c\\x16Trichomonascus_ciferii\\x94\\x8c\\x12Saprochaete_ingens\\x94\\x8c\\x15Saprochaete_fungicola\\x94\\x8c\\x13Nadsonia_fulvescens\\x94\\x8c\\x19Nadsonia_starkeyihenricii\\x94\\x8c\\x17Middelhovenomyces_tepae\\x94\\x8c\\x17Magnusiomyces_capitatus\\x94\\x8c\\x1aMagnusiomyces_tetraspermus\\x94\\x8c\\x12Dipodascus_albidus\\x94\\x8c\\x1eGroenewaldozyma_salmanticensis\\x94\\x8c\\x14Zygoascus_ofunaensis\\x94\\x8c\\x11Zygoascus_meyerae\\x94\\x8c\\x1dDiddensiella_caesifluorescens\\x94\\x8c\\x1cSpencermartinsiella_europaea\\x94\\x8c\\x16Blastobotrys_muscicola\\x94\\x8c\\x16Blastobotrys_mokoenaii\\x94\\x8c\\x16Blastobotrys_aristatus\\x94\\x8c Blastobotrys_raffinosifermentans\\x94\\x8c\\x1aSugiyamaella_lignohabitans\\x94\\x8c\\x17Sugiyamaella_xylanicola\\x94e\\x8c\\x10Alloascoideaceae\\x94]\\x94(\\x8c\\x16Alloascoidea_hylecoeti\\x94\\x8c\\x19Sporopachydermia_quercuum\\x94\\x8c\\x1cSporopachydermia_lactativora\\x94e\\x8c\\x07CUG-Ala\\x94]\\x94(\\x8c\\x16Pachysolen_tannophilus\\x94\\x8c\\x11Peterozyma_xylosa\\x94\\x8c\\x12Nakazawaea_peltata\\x94\\x8c\\x12Nakazawaea_holstii\\x94e\\x8c\\nPichiaceae\\x94]\\x94(\\x8c\\x13Kuraishia_capsulata\\x94\\x8c\\x10Kuraishia_ogatae\\x94\\x8c\\x10Candida_boidinii\\x94\\x8c\\x19Candida_arabinofermentans\\x94\\x8c\\x16Ogataea_parapolymorpha\\x94\\x8c\\x12Ogataea_polymorpha\\x94\\x8c\\x13Ogataea_methylovora\\x94\\x8c\\x18Ogataea_trehaloabstinens\\x94\\x8c\\x0fOgataea_kodamae\\x94\\x8c\\x0eOgataea_minuta\\x94\\x8c\\x10Ogataea_henricii\\x94\\x8c\\x0fOgataea_zsoltii\\x94\\x8c\\x12Ogataea_pilisensis\\x94\\x8c\\x12Ogataea_naganishii\\x94\\x8c\\x13Ogataea_ramenticola\\x94\\x8c\\x14Ambrosiozyma_maleeae\\x94\\x8c Ambrosiozyma_pseudovanderkliftii\\x94\\x8c\\x17Ambrosiozyma_philentoma\\x94\\x8c\\x1aAmbrosiozyma_kashinagicola\\x94\\x8c\\x14Dekkera_bruxellensis\\x94\\x8c\\x13Brettanomyces_nanus\\x94\\x8c\\x1aBrettanomyces_naardenensis\\x94\\x8c\\x18Kregervanrija_delftensis\\x94\\x8c\\x15Kregervanrija_fluxuum\\x94\\x8c\\x18Martiniozyma_abietophila\\x94\\x8c\\x13Saturnispora_silvae\\x94\\x8c\\x1cSaturnispora_serradocipensis\\x94\\x8c\\x16Saturnispora_zaruensis\\x94\\x8c\\x16Saturnispora_mendoncae\\x94\\x8c\\x14Saturnispora_dispora\\x94\\x8c\\x16Pichia_membranifaciens\\x94\\x8c\\x13Candida_californica\\x94\\x8c\\x13Pichia_kudriavzevii\\x94\\x8c\\x10Pichia_terricola\\x94\\x8c\\x12Pichia_norvegensis\\x94\\x8c\\x0ePichia_nakasei\\x94\\x8c\\rPichia_exigua\\x94\\x8c\\x11Pichia_manshurica\\x94\\x8c\\x13Pichia_occidentalis\\x94\\x8c\\x13Candida_inconspicua\\x94\\x8c\\x17Komagataella_kurtzmanii\\x94\\x8c\\x14Komagataella_phaffii\\x94\\x8c\\x15Komagataella_pastoris\\x94\\x8c\\x15Citeromyces_siamensis\\x94\\x8c\\x17Citeromyces_matritensis\\x94e\\x8c\\x08CUG-Ser2\\x94]\\x94(\\x8c\\x1bSaccharomycopsis_fibuligera\\x94\\x8c\\x18Saccharomycopsis_malanga\\x94\\x8c\\x1bSaccharomycopsis_fermentans\\x94\\x8c\\x18Saccharomycopsis_fodiens\\x94\\x8c\\x12Ascoidea_rubescens\\x94\\x8c\\x11Ascoidea_asiatica\\x94e\\x8c\\x08CUG-Ser1\\x94]\\x94(\\x8c\\x17Babjeviella_inositovora\\x94\\x8c\\x15Cephaloascus_fragrans\\x94\\x8c\\x14Cephaloascus_albidus\\x94\\x8c\\x0eDiutina_rugosa\\x94\\x8c\\x12Diutina_catenulata\\x94\\x8c\\x11Candida_schatavii\\x94\\x8c\\x18Kurtzmaniella_cleridarum\\x94\\x8c\\x13Kurtzmaniella_fragi\\x94\\x8c\\x16Candida_ascalaphidarum\\x94\\x8c\\x10Candida_gorgasii\\x94\\x8c\\x19Scheffersomyces_spartinae\\x94\\x8c\\x10Candida_jaroonii\\x94\\x8c\\x11Yamadazyma_tenuis\\x94\\x8c\\x14Yamadazyma_philogaea\\x94\\x8c\\x12Yamadazyma_scolyti\\x94\\x8c\\x13Candida_railenensis\\x94\\x8c\\x13Millerozyma_acaciae\\x94\\x8c\\x14Millerozyma_farinosa\\x94\\x8c\\x13Debaryomyces_fabryi\\x94\\x8c\\x14Debaryomyces_hanseni\\x94\\x8c\\x14Debaryomyces_maramus\\x94\\x8c\\x14Priceomyces_carsonii\\x94\\x8c\\x12Priceomyces_medius\\x94\\x8c\\x15Priceomyces_castillae\\x94\\x8c\\x17Wickerhamia_fluorescens\\x94\\x8c\\x15Teunomyces_gatunensis\\x94\\x8c\\x12Teunomyces_kruisii\\x94\\x8c\\x13Suhomyces_pyralidae\\x94\\x8c\\x17Suhomyces_canberraensis\\x94\\x8c\\x12Kodamaea_laetipori\\x94\\x8c\\x18Scheffersomyces_lignosus\\x94\\x8c\\x18Danielozyma_ontarioensis\\x94\\x8c\\x14Hyphopichia_burtonii\\x94\\x8c\\x17Hyphopichia_homilentoma\\x94\\x8c\\x11Hyphopichia_gotoi\\x94\\x8c\\x19Meyerozyma_guilliermondii\\x94\\x8c\\x10Meyerozyma_spJA9\\x94\\x8c\\x14Meyerozyma_caribbica\\x94\\x8c\\x16Meyerozyma_athensensis\\x94\\x8c\\rCandida_auris\\x94\\x8c\\x11Candida_haemuloni\\x94\\x8c\\x18Candida_pseudohaemulonii\\x94\\x8c\\x18Candida_duobushaemulonis\\x94\\x8c\\x18Metschnikowia_spJCM33374\\x94\\x8c\\x19Metschnikowia_bicuspidata\\x94\\x8c\\x19Metschnikowia_pulcherrima\\x94\\x8c\\x12Candida_intermedia\\x94\\x8c\\x15Clavispora_lusitaniae\\x94\\x8c\\x16Suhomyces_tanzawaensis\\x94\\x8c\\x18Scheffersomyces_stipitis\\x94\\x8c\\x11Candida_subhashii\\x94\\x8c\\x18Spathaspora_passalidarum\\x94\\x8c\\x11Spathaspora_spJA1\\x94\\x8c\\x14Candida_dubliniensis\\x94\\x8c\\x10Candida_albicans\\x94\\x8c\\x11Candida_tropcalis\\x94\\x8c\\x0fCandida_maltosa\\x94\\x8c\\x13Candida_oxycetoniae\\x94\\x8c\\x19Lodderomyces_elongesporus\\x94\\x8c\\x14Candida_jiufengensis\\x94\\x8c\\rCandida_theae\\x94\\x8c\\x14Candida_metapsilosis\\x94\\x8c\\x14Candida_parapsilosis\\x94\\x8c\\x15Candida_orthopsilosis\\x94\\x8c\\x11Candida_chilensis\\x94e\\x8c\\x12Saccharomycetaceae\\x94]\\x94(\\x8c\\x16Tetrapisispora_blattae\\x94\\x8c\\x19Vanderwaltozyma_polyspora\\x94\\x8c\\x16Tetrapisispora_phaffii\\x94\\x8c\\x18Zygosaccharomyces_mellis\\x94\\x8c\\x18Zygosaccharomyces_rouxii\\x94\\x8c\\x18Zygosaccharomyces_bailii\\x94\\x8c\\x1cZygosaccharomyces_parabailii\\x94\\x8c\\x16Zygotorulaspora_mrakii\\x94\\x8c\\x13Torulaspora_globosa\\x94\\x8c\\x15Torulaspora_spCBS2947\\x94\\x8c\\x17Torulaspora_delbrueckii\\x94\\x8c\\x10Candida_glabrata\\x94\\x8c\\x17Saccharomyces_paradoxus\\x94\\x8c\\x19Saccharomyces_pastorianus\\x94\\x8c\\x17Saccharomyces_eubayanus\\x94\\x8c\\x17Saccharomyces_boulardii\\x94\\x8c\\x18Saccharomyces_cerevisiae\\x94\\x8c\\x15Naumovozyma_castellii\\x94\\x8c\\x17Naumovozyma_dairenensis\\x94\\x8c\\x1aNakaseomyces_bacillisporus\\x94\\x8c\\x15Kazachstania_africana\\x94\\x8c\\x13Kazachstania_exigua\\x94\\x8c\\x18Kazachstania_saulgeensis\\x94\\x8c\\x16Kazachstania_barnettii\\x94\\x8c\\x17Kazachstania_naganishii\\x94\\x8c\\x15Kazachstania_unispora\\x94\\x8c\\x17Eremothecium_sinecaudum\\x94\\x8c\\x18Eremothecium_cymbalariae\\x94\\x8c\\x14Eremothecium_gosypii\\x94\\x8c\\x0cAshbya_aceri\\x94\\x8c\\x17Kluyveromyces_marxianus\\x94\\x8c\\x13Kluveromyces_lactis\\x94\\x8c\\x19Kluyveromyces_dobzhanskii\\x94\\x8c\\x13Lachancea_mirantina\\x94\\x8c\\x14Lachancea_fermentati\\x94\\x8c\\x15Lachancea_quebecensis\\x94\\x8c\\x18Lachancea_thermotolerans\\x94\\x8c\\x13Lachancea_nothofagi\\x94\\x8c\\x13Lachancea_dasiensis\\x94\\x8c\\x17Lachancea_lanzerotensis\\x94\\x8c\\x12Lachancea_meyersii\\x94\\x8c\\x12Yueomyces_sinensis\\x94\\x8c\\x19Grigorovia_yakushimaensis\\x94\\x8c\\x19Grigorovia_transvaalensis\\x94e\\x8c\\x12Saccharomycodaceae\\x94]\\x94(\\x8c\\x17Saccharomycodes_ludwigi\\x94\\x8c\\x17Hanseniaspora_osmophila\\x94\\x8c\\x1cHanseniaspora_guilliermondii\\x94\\x8c\\x16Hanseniaspora_opuntiae\\x94\\x8c\\x14Hanseniaspora_uvarum\\x94e\\x8c\\x10Phaffomycetaceae\\x94]\\x94(\\x8c\\x12Candida_ponderosae\\x94\\x8c\\x11Starmera_quercuum\\x94\\x8c\\x14Starmera_amethionina\\x94\\x8c\\x17Wickerhamomyces_pijperi\\x94\\x8c\\x17Wickerhamomyces_mucosus\\x94\\x8c\\x18Wickerhamomyces_ciferrii\\x94\\x8c\\x18Wickerhamomyces_anomalus\\x94\\x8c\\x14Wickerhamomyces_alni\\x94\\x8c\\x15Wickerhamomyces_bovis\\x94\\x8c\\x15Cyberlindnera_jadinii\\x94\\x8c\\x16Cyberlindnera_fabianii\\x94\\x8c\\x14Cyberlindnera_mrakii\\x94\\x8c\\x18Barnettozyma_californica\\x94\\x8c\\x16Barnettozyma_salicaria\\x94\\x8c\\x16Barnettozyma_pratensis\\x94\\x8c\\x0fCandida_montana\\x94\\x8c\\x14Phaffomyces_opuntiae\\x94euu\\x8c\\x04rule\\x94\\x8c\\x0fdraw_phenograms\\x94\\x8c\\x0fbench_iteration\\x94N\\x8c\\tscriptdir\\x94\\x8cg/cluster/ggs_lab/mtparker/papers/splice_site_phylogenetics/phylo_assoc_mapping/rules/notebook_templates\\x94ub.'); from snakemake.logging import logger; logger.printshellcmds = False; import os; os.chdir(r'/cluster/ggs_lab/mtparker/papers/splice_site_phylogenetics/phylo_assoc_mapping/pipeline/ascomycota');\n",
    "######## snakemake preamble end #########\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "from glob import glob\n",
    "import tempfile\n",
    "import warnings\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import ListedColormap, LinearSegmentedColormap, Normalize, to_hex\n",
    "from matplotlib.cm import ScalarMappable\n",
    "import seaborn as sns\n",
    "\n",
    "import ete3\n",
    "os.environ['QT_QPA_PLATFORM']='offscreen'\n",
    "\n",
    "from rpy2 import robjects as robj\n",
    "from rpy2.robjects.packages import importr, isinstalled\n",
    "from rpy2.robjects.conversion import localconverter\n",
    "from rpy2.rinterface import RRuntimeWarning\n",
    "from rpy2.rinterface_lib.embedded import RRuntimeError\n",
    "\n",
    "from rpy2.robjects import pandas2ri\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=RRuntimeWarning)\n",
    "\n",
    "\n",
    "def import_or_install(package_name, repo='https://cran.ma.imperial.ac.uk/'):\n",
    "    if not isinstalled(package_name):\n",
    "        utils = importr('utils')\n",
    "        utils.install_packages(robj.StrVector([package_name]), repos=repo)\n",
    "    return importr(package_name)\n",
    "\n",
    "\n",
    "ape = import_or_install('ape')\n",
    "\n",
    "## Default plotting params\n",
    "%matplotlib inline\n",
    "sns.set(font='Arial')\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "style = sns.axes_style('white')\n",
    "style.update(sns.axes_style('ticks'))\n",
    "style['xtick.major.size'] = 2\n",
    "style['ytick.major.size'] = 2\n",
    "sns.set(font_scale=1.2, style=style)\n",
    "pal = sns.color_palette(['#0072b2', '#d55e00', '#009e73', '#f0e442', '#cc79a7', '#56b4e9', '#e69f00'])\n",
    "cmap = ListedColormap(pal.as_hex())\n",
    "sns.set_palette(pal)\n",
    "sns.palplot(pal)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sign, pos, nom_base, denom_base = re.search(\n",
    "    '([pm])(\\d+)([ACGURYSKMWBDHVN])to([ACGURYSKMWBDHVN])',\n",
    "    snakemake.wildcards.phenotype\n",
    ").groups()\n",
    "\n",
    "abspos = int(pos) - 1 if sign == 'p' else -int(pos)\n",
    "if snakemake.wildcards.signal_type == '5ss':\n",
    "    abspos = 3 + abspos\n",
    "print(abspos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = ete3.Tree(snakemake.input.tree, format=1)\n",
    "utree = tree.copy()\n",
    "utree.convert_to_ultrametric(tree_length=1)\n",
    "species = [n.name for n in tree.get_leaves()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "IUPAC = {\n",
    "    'A': ['A'],\n",
    "    'C': ['C'],\n",
    "    'G': ['G'],\n",
    "    'U': ['U'],\n",
    "    'R': ['A', 'G'],\n",
    "    'Y': ['C', 'U'],\n",
    "    'S': ['G', 'C'],\n",
    "    'W': ['A', 'U'],\n",
    "    'K': ['G', 'U'],\n",
    "    'M': ['A', 'C'],\n",
    "    'B': ['C', 'G', 'U'],\n",
    "    'D': ['A', 'G', 'U'],\n",
    "    'H': ['A', 'C', 'U'],\n",
    "    'V': ['A', 'C', 'G'],\n",
    "    'N': ['A', 'C', 'G', 'U']\n",
    "}\n",
    "\n",
    "pssm_dir = os.path.split(snakemake.input.pssms[0])[0]\n",
    "pssm_fns = glob(f'{pssm_dir}/*.{snakemake.wildcards.signal_type}.tsv')\n",
    "\n",
    "pssms = {\n",
    "    os.path.split(fn)[1].split('.')[0]: pd.read_csv(fn, sep='\\t', index_col=[0, 1])\n",
    "    for fn in pssm_fns\n",
    "}\n",
    "\n",
    "bar_pheno = {}\n",
    "pheno = {}\n",
    "for node, pssm in pssms.items():\n",
    "    mean_pssm = pssm.loc[pd.IndexSlice[:, abspos], :].mean(0)\n",
    "    # sankoff parsimony phenos used for stacked bar charts\n",
    "    bar_pheno[node] = mean_pssm\n",
    "    if node in species:\n",
    "        # internal phenotypes for phenograms will be calculated with APE so only take pheno for leaves\n",
    "        pheno[node] = mean_pssm[IUPAC[nom_base]].sum() - mean_pssm[IUPAC[denom_base]].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_ape_tree(tree):\n",
    "    tmp_fh, tmp_fn = tempfile.mkstemp()\n",
    "    tree.write(outfile=tmp_fn, format=1, format_root_node=True)\n",
    "    r_tree = ape.read_tree(tmp_fn)\n",
    "    os.close(tmp_fh)\n",
    "    os.remove(tmp_fn)\n",
    "    return r_tree\n",
    "\n",
    "\n",
    "def get_ancestral_states(tree, pheno):\n",
    "    r_tree = to_ape_tree(tree)\n",
    "    leaf_order = list(r_tree.rx2['tip.label'])\n",
    "    pheno = [pheno[n] for n in leaf_order]\n",
    "    anc = ape.ace(\n",
    "        robj.FloatVector(pheno),\n",
    "        r_tree,\n",
    "        type='continuous'\n",
    "    )\n",
    "    return dict(zip(r_tree.rx2['node.label'], anc.rx2['ace']))\n",
    "\n",
    "\n",
    "def _plot_recursive(node, pheno, x1, y1, colour, ax, clade_root_node_colours, label_species):\n",
    "    for child in node.children:\n",
    "        x2 = x1 + child.dist\n",
    "        y2 = pheno[child.name]\n",
    "        node_leaves = set([n.name for n in child.get_leaves()])\n",
    "        if node_leaves.intersection(label_species):\n",
    "            lw = 3\n",
    "            zorder = 1\n",
    "            alpha = 1\n",
    "        else:\n",
    "            lw = 0.8\n",
    "            zorder = 0\n",
    "            alpha = 0.75\n",
    "        ax.plot([x1, x2], [y1, y2], color=colour, lw=lw, zorder=zorder, alpha=alpha)\n",
    "        if not child.is_leaf():\n",
    "            if child.name in clade_root_node_colours:\n",
    "                _plot_recursive(child, pheno, x2, y2,\n",
    "                                clade_root_node_colours[child.name],\n",
    "                                ax, clade_root_node_colours,\n",
    "                                label_species)\n",
    "            else:\n",
    "                _plot_recursive(child, pheno, x2, y2, colour, ax,\n",
    "                                clade_root_node_colours, label_species)\n",
    "        else:\n",
    "            if child.name in label_species:\n",
    "                ax.annotate(xy=(x2, y2), s=child.name,\n",
    "                            ha='left', va='center', color=colour)\n",
    "\n",
    "    \n",
    "def plot_stacked_pheno_histogram(pheno, species_clades, clade_colours, ax):\n",
    "    hist_data = []\n",
    "    colours = []\n",
    "    labels = []\n",
    "    for clade_name, species_list in species_clades.items():\n",
    "        hist_data.append([pheno[sp] for sp in species_list])\n",
    "        colours.append(clade_colours[clade_name])\n",
    "        labels.append(clade_name)\n",
    "\n",
    "    for i in range(len(hist_data)):\n",
    "        ax.hist(\n",
    "            np.concatenate(hist_data[i:]),\n",
    "            orientation='horizontal',\n",
    "            range=(-1, 1),\n",
    "            bins=20,\n",
    "            color=colours[i],\n",
    "            label=labels[i],\n",
    "            edgecolor='none',\n",
    "        )\n",
    "    ax.legend(title='Clades')\n",
    "    return ax\n",
    "\n",
    "\n",
    "def plot_phenogram(tree, pheno,\n",
    "                   species_clades, clade_colours,\n",
    "                   title=None, ultrametric=True,\n",
    "                   label_species=None, figsize=(15, 10), add_hist=True):\n",
    "    tree = tree.copy()\n",
    "    if ultrametric:\n",
    "        tree.convert_to_ultrametric(tree_length=1)\n",
    "    anc_states = get_ancestral_states(tree, pheno)\n",
    "    anc_states.update(pheno)\n",
    "\n",
    "    clade_root_node_colours = {}\n",
    "\n",
    "    for clade_name, species_list in species_clades.items():\n",
    "        clade_root_node_colours[\n",
    "            tree.get_common_ancestor(species_list).name\n",
    "        ] = clade_colours[clade_name]\n",
    "\n",
    "    fig = plt.figure(figsize=figsize)\n",
    "    if add_hist:\n",
    "        phenogram_ax = plt.subplot2grid((1, 3), (0, 0), colspan=2)\n",
    "        hist_ax = plt.subplot2grid((1, 3), (0, 2), colspan=1, sharey=phenogram_ax)\n",
    "        plot_stacked_pheno_histogram(pheno, species_clades, clade_colours, hist_ax)\n",
    "        hist_ax.set_xlabel('Species')\n",
    "    else:\n",
    "        phenogram_ax = plt.subplots()\n",
    "        hist_ax = None\n",
    "    if label_species is None:\n",
    "        label_species = []\n",
    "    x1 = 0\n",
    "    y2 = anc_states[tree.name]\n",
    "    _plot_recursive(\n",
    "        tree, anc_states, x1, y2,\n",
    "        clade_root_node_colours[tree.name],\n",
    "        phenogram_ax,\n",
    "        clade_root_node_colours,\n",
    "        label_species)\n",
    "    if ultrametric:\n",
    "        phenogram_ax.set_xlabel('Branch length (ultrametric)')\n",
    "    else:\n",
    "        phenogram_ax.set_xlabel('Branch length')\n",
    "    if title is not None:\n",
    "        phenogram_ax.set_ylabel(title)\n",
    "    plt.tight_layout()\n",
    "    return fig, phenogram_ax, hist_ax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_name = {'p': '+', 'm': '-'}[sign] + pos\n",
    "title = f'{pos_name} {nom_base} to {denom_base} ratio'\n",
    "\n",
    "clade_colours = {\n",
    "    'Outgroups': '#7c7c7c',\n",
    "    'Lipomycetaceae': '#f1c232',\n",
    "    'Trigonopsidaceae': '#bf9000',\n",
    "    'Dipodascaceae': '#ff0000',\n",
    "    'Alloascoideaceae': '#c27ba0',\n",
    "    'CUG-Ser2': '#d55e00',\n",
    "    'Phaffomycetaceae': '#ee9956',\n",
    "    'CUG-Ser1': '#0072b2',\n",
    "    'CUG-Ala': '#56b4e9',\n",
    "    'Pichiaceae': '#674ea7',\n",
    "    'Saccharomycodaceae': '#00f3b1',\n",
    "    'Saccharomycetaceae': '#009e73',\n",
    "}\n",
    "\n",
    "\n",
    "fig, p_ax, h_ax = plot_phenogram(\n",
    "    tree, pheno,\n",
    "    snakemake.config['known_clades'],\n",
    "    clade_colours,\n",
    "    title=title,\n",
    "    label_species=['Homo_sapiens',\n",
    "                   'Candida_albicans',\n",
    "                   'Yarrowia_lipolytica',\n",
    "                  # 'Lipomyces_starkeyi',\n",
    "                   'Saccharomyces_cerevisiae',\n",
    "                  # 'Eremothecium_gosypii',\n",
    "                 #  'Cephaloascus_fragrans',\n",
    "                   'Schizosaccharomyces_pombe',\n",
    "                   'Crytococcus_neoformans']\n",
    ")\n",
    "\n",
    "p_ax.set_ylim(-1, 1)\n",
    "plt.savefig(snakemake.output.phenogram)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "IUPAC_COLORS = [pal.as_hex()[2], pal.as_hex()[0], pal.as_hex()[3], pal.as_hex()[1]]\n",
    "\n",
    "def draw_collapsed_tree_with_bars(tree, bar_data, species_clades, reference_species, clade_colours):\n",
    "    \n",
    "    t = tree.copy()\n",
    "\n",
    "    keep_nodes = set(reference_species)\n",
    "    for sp in reference_species:\n",
    "        bd = bar_data[sp]\n",
    "        bd = bd.values / bd.sum() * 100\n",
    "        barface = ete3.StackedBarFace(bd, 150, 15, colors=IUPAC_COLORS)\n",
    "        n = t.search_nodes(name=sp)[0]\n",
    "        n.add_face(barface, 0, 'aligned')\n",
    "        try:\n",
    "            n.img_style['bgcolor'] = clade_colours[species_clades[sp]]\n",
    "        except KeyError:\n",
    "            n.img_style['bgcolor'] = clade_colours['Outgroups']\n",
    "    for clade_name, species_list in species_clades.items():\n",
    "        lca = t.get_common_ancestor(species_list)\n",
    "        if lca.name != t.get_tree_root().name:\n",
    "            bd = bar_data[lca.name]\n",
    "            bd = bd.values / bd.sum() * 100\n",
    "            lca.name = clade_name\n",
    "            keep_nodes.add(clade_name)\n",
    "            keep_nodes.update([n.name for n in lca.get_ancestors()])\n",
    "            barface = ete3.StackedBarFace(bd, 150, 15, colors=IUPAC_COLORS)\n",
    "            lca.add_face(barface, 0, 'aligned')\n",
    "            lca.img_style['bgcolor'] = clade_colours[clade_name]\n",
    "\n",
    "    t.prune(keep_nodes)\n",
    "    t.convert_to_ultrametric()\n",
    "    t.ladderize()\n",
    "\n",
    "    cmap = LinearSegmentedColormap.from_list(name='redblack', colors=['#ff0000', '#252525'], N=100)\n",
    "    # colour branches by confidence\n",
    "    for node in t.traverse():\n",
    "        color = to_hex(cmap(float(node.conf)))\n",
    "        node.img_style['vt_line_color'] = color\n",
    "\n",
    "    ts = ete3.TreeStyle()\n",
    "    ts.show_leaf_name = True\n",
    "\n",
    "    return t, ts, cmap\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "t, ts, cmap = draw_collapsed_tree_with_bars(\n",
    "    tree, bar_pheno,\n",
    "    snakemake.config['known_clades'],\n",
    "    ['Homo_sapiens', 'Schizosaccharomyces_pombe', 'Crytococcus_neoformans'],\n",
    "    clade_colours\n",
    ")\n",
    "t.render(snakemake.output.collapsed_tree,\n",
    "         tree_style=ts)\n",
    "t.render(\"%%inline\", tree_style=ts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(0.75, 7))\n",
    "img = plt.imshow(np.array([[0, 1]]), cmap=cmap)\n",
    "plt.gca().set_visible(False)\n",
    "cax = plt.axes([0.1, 0.2, 0.8, 0.6])\n",
    "plt.colorbar(orientation=\"vertical\", cax=cax)\n",
    "cax.set_ylabel('Fraction of Orthofinder\\ngene trees supporting split')\n",
    "plt.savefig(snakemake.output.collapsed_tree_cbar)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_orthofinder_n0(n0_fn):\n",
    "    og = pd.read_csv(n0_fn, sep='\\t', index_col='HOG').fillna('')\n",
    "    og = og.iloc[:, 2:]\n",
    "    return og\n",
    "\n",
    "def get_key_ogs(key_prot_ids, og):\n",
    "    og_ids = []\n",
    "    for _, gene_symbol, prot_id, species_name in key_prot_ids.itertuples():\n",
    "        try:\n",
    "            og_id = og.query(f'{species_name}.str.contains(\"{prot_id}\")').index[0]\n",
    "        except IndexError:\n",
    "            warnings.warn(f'Could not find orthogroup for \"{gene_symbol}\"')\n",
    "            og_id = np.nan\n",
    "        og_ids.append(og_id)\n",
    "    key_prot_ids['og_id'] = og_ids\n",
    "    og_idx = []\n",
    "    new_og_ids = []\n",
    "    for og_id, group in key_prot_ids.groupby('og_id'):\n",
    "        og_idx.append(og_id)\n",
    "        new_og_ids.append('_'.join(group.gene_symbol))\n",
    "    og_key = og.loc[og_idx]\n",
    "    og_key.index = new_og_ids\n",
    "    return og_key\n",
    "\n",
    "\n",
    "key_og_n0 = read_orthofinder_n0(snakemake.input.key_ogs)\n",
    "key_prot_table = pd.read_csv(\n",
    "    snakemake.input.key_prot_ids, sep='\\t', names=['gene_symbol', 'prot_id', 'species']\n",
    ")\n",
    "key_og_n0 = get_key_ogs(key_prot_table, key_og_n0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw_full_circular_tree_with_bars(tree, bar_data, og_data, species_clades, clade_colours):\n",
    "    \n",
    "    t = tree.copy()\n",
    "\n",
    "    for node in t.get_leaves():\n",
    "        bd = bar_data[node.name]\n",
    "        bd = bd.values / bd.sum() * 100\n",
    "        for clade_name, sp_list in species_clades.items():\n",
    "            if node.name in sp_list:\n",
    "                colour = clade_colours[clade_name]\n",
    "                break\n",
    "        else:\n",
    "            colour = clade_colours['Outgroups']\n",
    "        genus, species = node.name.split('_', 1)\n",
    "        name = f'{genus[0]}. {species}'\n",
    "        label = ete3.RectFace(300, 25,\n",
    "                              label={'text': name, 'color': '#252525', 'fontsize':20, 'font': 'Arial'},\n",
    "                              bgcolor=colour, fgcolor=None)\n",
    "        node.add_face(label, 0, 'aligned')\n",
    "        # spacer\n",
    "        node.add_face(ete3.RectFace(10, 25, bgcolor=None, fgcolor=None), 1, 'aligned')\n",
    "        barface = ete3.StackedBarFace(bd, 300, 25, colors=IUPAC_COLORS)\n",
    "        barface.rotable = False\n",
    "        node.add_face(barface, 2, 'aligned')\n",
    "        i = 3\n",
    "        for og_name, ogd in og_data.items():\n",
    "            node.add_face(ete3.RectFace(10, 25, bgcolor=None, fgcolor=None), i, 'aligned')\n",
    "            i += 1\n",
    "            o = ete3.CircleFace(15, color='#252525' if ogd[node.name] else '#aaaaaa')\n",
    "            node.add_face(o, i, 'aligned')\n",
    "            i += 1\n",
    "\n",
    "    t.ladderize()\n",
    "\n",
    "    cmap = LinearSegmentedColormap.from_list(name='redblack', colors=['#ff0001', '#252526'], N=100)\n",
    "    # colour branches by confidence\n",
    "    for node in t.traverse():\n",
    "        color = to_hex(cmap(float(node.conf)))\n",
    "        node.img_style['vt_line_color'] = color\n",
    "        node.img_style['vt_line_width'] = 5\n",
    "        node.img_style['hz_line_width'] = 5\n",
    "        node.img_style[\"size\"]=0\n",
    "\n",
    "    ts = ete3.TreeStyle()\n",
    "    ts.show_leaf_name = False\n",
    "    ts.mode = 'c'\n",
    "    ts.arc_start = 5\n",
    "    ts.arc_span = 350\n",
    "\n",
    "    return t, ts, cmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "og_data = {\n",
    "    'METTL16': key_og_n0.loc['METTL16'].to_dict(),\n",
    "    'SNRNP27': key_og_n0.loc['SNRNP27'].to_dict(),\n",
    "}\n",
    "\n",
    "t, ts, cmap = draw_full_circular_tree_with_bars(\n",
    "    tree, bar_pheno, og_data,\n",
    "    snakemake.config['known_clades'],\n",
    "    clade_colours\n",
    ")\n",
    "t.render(snakemake.output.full_circular_tree, tree_style=ts);\n",
    "t.render(\"%%inline\", tree_style=ts)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
